test/c/test_sparse_observable.c

```diff
@@ -0,0 +1,650 @@
+int test_mult() {
+    complex double coeffs[3] = {2, 2 * I, 2 + 2 * I};
+
+    for (int i = 0; i < 3; i++) {
+        QkSparseObservable *obs = qk_obs_identity(100);
+
+        QkSparseObservable *result = qk_obs_multiply(obs, &coeffs[i]);
+
+        // construct the expected observable: coeff * Id
+        QkSparseObservable *expected = qk_obs_zero(100);
+        QkBitTerm bit_terms[] = {};
+        uint32_t indices[] = {};
+        QkSparseTerm term = {coeffs[i], 0, bit_terms, indices, 100};
+        qk_obs_add_term(expected, &term);
+
+        // perform the check
+        bool is_equal = qk_obs_equal(expected, result);
+
+        // deallocate before returning
+        qk_obs_free(obs);
+        qk_obs_free(result);
+        qk_obs_free(expected);
+
+        if (!is_equal) {
+            return EqualityError;
+        }
+    }
+
+    return Ok;
+}
+
+/**
+ * Test bringing an observable into canonical form.
+ */
+int test_canonicalize() {
+    QkSparseObservable *left = qk_obs_identity(100);
+    QkSparseObservable *right = qk_obs_identity(100);
+    QkSparseObservable *obs = qk_obs_add(left, right);
+
+    double tol = 1e-5;
+    QkSparseObservable *simplified = qk_obs_canonicalize(obs, tol);
+
+    // construct the expected observable: 2 * Id
+    QkSparseObservable *expected = qk_obs_zero(100);
+    QkBitTerm bit_terms[] = {};
+    uint32_t indices[] = {};
+    complex double coeff = 2.0;
+    QkSparseTerm term = {coeff, 0, bit_terms, indices, 100};
+    qk_obs_add_term(expected, &term);
+
+    bool is_equal = qk_obs_equal(expected, simplified);
+
+    qk_obs_free(obs);
+    qk_obs_free(right);
+    qk_obs_free(left);
+    qk_obs_free(simplified);
+    qk_obs_free(expected);
+
+    if (!is_equal) {
+        return EqualityError;
+    }
+
+    return Ok;
+}
+
+/**
+ * Test getting the number of terms in an observable.
+ */
+int test_num_terms() {
+    int result = Ok;
+    uint64_t num_terms;
+
+    QkSparseObservable *zero = qk_obs_zero(100);
+    num_terms = qk_obs_num_terms(zero);
+    if (num_terms != 0) {
+        result = EqualityError;
+    }
+    qk_obs_free(zero);
+
+    QkSparseObservable *iden = qk_obs_identity(100);
+    num_terms = qk_obs_num_terms(iden);
+    if (num_terms != 1) {
+        result = EqualityError;
+    }
+    qk_obs_free(iden);
+
+    return result;
+}
+
+/**
+ * Test getting the number of qubits in an observable.
+ */
+int test_num_qubits() {
+    int result = Ok;
+    uint32_t num_qubits;
+
+    QkSparseObservable *obs = qk_obs_zero(1);
+    num_qubits = qk_obs_num_qubits(obs);
+    if (num_qubits != 1) {
+        result = EqualityError;
+    }
+    qk_obs_free(obs);
+
+    QkSparseObservable *obs100 = qk_obs_zero(100);
+    num_qubits = qk_obs_num_qubits(obs100);
+    if (num_qubits != 100) {
+        result = EqualityError;
+    }
+    qk_obs_free(obs100);
+
+    return result;
+}
+
+/**
+ * Test adding an individual term to an observable.
+ */
+int test_custom_build() {
+    u_int32_t num_qubits = 100;
```

Comment:
Pretty sure that `u_int32_t` is some weird BSDish extension.  I think we might just be getting lucky that Apple clang and the Ubuntu gcc support it as well - I don't think it's a required part of `stdint.h`.

Reply:
I fixed this in one of my commits.